src/mscorlib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -556,25 +761,31 @@ internal void UpdateValue(int entryIndex, TValue newValue)
             //----------------------------------------------------------------------------------------
             internal Container Resize()
             {
-                // Start by assuming we won't resize.
-                int newSize = _buckets.Length;
+                Debug.Assert(!HasCapacity);
 
-                // If any expired or removed keys exist, we won't resize.
                 bool hasExpiredEntries = false;
-                for (int entriesIndex = 0; entriesIndex < _entries.Length; entriesIndex++)
+                int newSize = _buckets.Length;
+
+                if (_parent == null || _parent._activeEnumeratorRefCount == 0)
                 {
-                    if (_entries[entriesIndex].HashCode == -1)
+                    // If any expired or removed keys exist, we won't resize.
+                    // If there any active enumerators, though, we don't want
+                    // to compact and thus have no expired entries.
+                    for (int entriesIndex = 0; entriesIndex < _entries.Length; entriesIndex++)
                     {
-                        // the entry was removed
-                        hasExpiredEntries = true;
-                        break;
-                    }
+                        if (_entries[entriesIndex].HashCode == -1)
```

Comment:
Can use `ref Entry entry` here to avoid redundant array lookups? (in case the JIT is not smart enough to eliminate them).